File: lxb/leetcode/MergeKSortedList.cpp
http://oj.leetcode.com/problems/merge-k-sorted-lists/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    struct Node{
        int index;
        ListNode *ln;
        
        Node(int i,ListNode* l):index(i),ln(l){};
    };
    
    struct compare{
        bool operator()(const Node &l, const Node &r){
            return l.ln->val > r.ln->val;
        }
    };
    
    ListNode *mergeKLists(vector<ListNode *> &lists) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        priority_queue<Node, vector<Node>, compare>   PQ;
        
        // init PQ
        for (int i=0; i < lists.size(); i++){
            if (lists[i])   {
                ListNode *tmp = lists[i];
                lists[i] = lists[i]->next;
                tmp->next = NULL;
                PQ.push(Node(i, tmp));
            }
        }
        
        // get from Q, insert into merged list. move original list to next;

        ListNode *result = NULL;
        ListNode *tail = NULL;
        while (!PQ.empty())
        {
            //get one from PQ
            ListNode *node=PQ.top().ln;
            int nextListIndex = PQ.top().index;
            PQ.pop();
        
            //insert it into result.
            if (tail == NULL){
                tail = node;
                result = tail;
            }else{
                tail->next = node;
                tail = node;
            }
        
            //refill one into PQ 
            if (lists[nextListIndex]){
                ListNode *tmp = lists[nextListIndex];
                lists[nextListIndex] = lists[nextListIndex]->next;
                tmp->next = NULL;
            
                //insert into PQ
                PQ.push(Node(nextListIndex, tmp));
            }
        }
        
        return result;
    }
};

File: lxb/leetcode/java/MergeKSortedList.java
/*
Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.
*/

Focus: 1. PriorityQueue usage. import java.util.PriorityQueue
	2. PriorityQueue<T> heap = new PriorityQueue<T>(arrlist.size(), comp)
	   Comparator<T> comp = new Comparator<T>() {
		int comparator(T o1, T o2) {
			...
			}
		}
	    heap.add(v) heap.peek(v) (vs c++ heap.top()), heap.poll() (vs c++ heap.pop()), heap.isEmpty()
	3. Algorithm. Maintain one heap and choose one and only one element from each list.
	4. Then get least one from heap, refill from that list. <Tip>: The next element

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode mergeKLists(ArrayList<ListNode> lists) {
        if (lists == null || lists.isEmpty()) {
            return null;
        }  
        
        Comparator<ListNode> comp = new Comparator<ListNode>() {
            public int compare(ListNode o1, ListNode o2) {
                if (o1.val < o2.val) {
                    return -1;
                } else if (o1.val > o2.val) {
                    return 1;
                } else {
                    return 0;
                }
            }
        } ;
        
        PriorityQueue<ListNode> heap = new PriorityQueue<ListNode>(lists.size(), comp);
        for (ListNode node: lists) {
            if (node != null) {
                heap.add(node);
            }
        }
        
        ListNode head = null;
        ListNode curr = null;
        
        while (!heap.isEmpty()) {
            if (head == null) {
                head = heap.peek();
                curr = heap.poll();
            } else {
                curr.next = heap.poll();
                curr = curr.next;
            }
            
            //refill one
            if (curr.next != null) {
                heap.add(curr.next);
            }
        }
        
        return head;
    }
}
